File: sdks/python/starfish/clock.py
from __future__ import annotations

import asyncio
import time
from typing import TYPE_CHECKING, Callable

from .id import next_id
from .types import StarfishFrame

if TYPE_CHECKING:
    from .connection import Connection

DEFAULT_SAMPLE_COUNT = 5


class Clock:
    def __init__(self, connection: Connection) -> None:
        self._connection = connection
        self._offset: float = 0

    @property
    def offset(self) -> float:
        return self._offset

    def now(self) -> float:
        return time.time() * 1000 + self._offset

    async def sync(self, samples: int = DEFAULT_SAMPLE_COUNT) -> float:
        offsets: list[float] = []

        for _ in range(samples):
            t1 = time.time() * 1000
            frame = StarfishFrame(v=1, id=next_id("clock"), type="clock.sync", ts=int(t1))
            response = await self._connection.send_and_wait(frame)
            t4 = time.time() * 1000

            server_time = response.payload.get("serverTime") if response.payload else None
            if server_time:
                rtt = t4 - t1
                estimated_server_time = server_time + rtt / 2
                offsets.append(estimated_server_time - t4)

        if offsets:
            offsets.sort()
            self._offset = offsets[len(offsets) // 2]

        return self._offset

    def at(self, server_time: float, callback: Callable[[], None]) -> asyncio.TimerHandle:
        local_time = server_time - self._offset
        delay = max(0, local_time - time.time() * 1000) / 1000
        loop = asyncio.get_running_loop()
        return loop.call_later(delay, callback)

File: sdks/golang/starfish/clock.go
package starfish

import (
	"context"
	"fmt"
	"sort"
	"sync"
	"time"
)

const defaultClockSamples = 5

// clockManager handles clock synchronization with the server.
type clockManager struct {
	mu     sync.RWMutex
	conn   *connection
	idg    *IDGenerator
	offset int64 // estimated server-client offset in ms
	synced bool
}

func newClockManager(conn *connection, idg *IDGenerator) *clockManager {
	return &clockManager{
		conn: conn,
		idg:  idg,
	}
}

// sync performs clock synchronization with the server using multiple samples.
func (c *clockManager) sync(ctx context.Context, samples int) error {
	if samples <= 0 {
		samples = defaultClockSamples
	}
	if samples < 3 {
		samples = 3 // minimum 3 per protocol spec
	}

	offsets := make([]int64, 0, samples)

	for i := 0; i < samples; i++ {
		offset, err := c.takeSample(ctx)
		if err != nil {
			return fmt.Errorf("starfish: clock sync sample %d failed: %w", i, err)
		}
		offsets = append(offsets, offset)

		// Small delay between samples to avoid burst
		if i < samples-1 {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(50 * time.Millisecond):
			}
		}
	}

	// Use median offset
	sort.Slice(offsets, func(i, j int) bool { return offsets[i] < offsets[j] })
	median := offsets[len(offsets)/2]

	c.mu.Lock()
	c.offset = median
	c.synced = true
	c.mu.Unlock()

	return nil
}

func (c *clockManager) takeSample(ctx context.Context) (int64, error) {
	sendTime := time.Now().UnixMilli()
	ts := sendTime

	frame := &Frame{
		Header: Header{
			V:        1,
			ID:       c.idg.Next("clock"),
			Resource: "clock",
			Method:   "sync",
			Kind:     "request",
			Ts:       &ts,
		},
	}

	reply, err := c.conn.sendAndWait(ctx, frame, 5*time.Second)
	if err != nil {
		return 0, err
	}

	receiveTime := time.Now().UnixMilli()
	rtt := receiveTime - sendTime

	serverTime, ok := reply.Payload["serverTime"].(float64)
	if !ok {
		return 0, fmt.Errorf("starfish: clock.synced missing serverTime")
	}

	// NTP-style offset: serverTime - (sendTime + rtt/2)
	offset := int64(serverTime) - (sendTime + rtt/2)
	return offset, nil
}

// now returns the estimated server time.
func (c *clockManager) now() int64 {
	c.mu.RLock()
	offset := c.offset
	c.mu.RUnlock()
	return time.Now().UnixMilli() + offset
}

// getOffset returns the current clock offset in milliseconds.
func (c *clockManager) getOffset() int64 {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.offset
}

// setInitialOffset sets the offset from the welcome message's serverTime.
func (c *clockManager) setInitialOffset(serverTime int64) {
	c.mu.Lock()
	c.offset = serverTime - time.Now().UnixMilli()
	c.mu.Unlock()
}
